subsys/senss/include/senss_sensor.h

```diff
@@ -0,0 +1,595 @@
+extern "C" {
+#endif
+
+
+/**
+ * @struct senss_sensor_reporter
+ * @brief Sensor reporter information for sensor register
+ *
+ * Can use this data structure to configure reporters' information when register
+ * a virtual sensor instance.
+ */
+struct senss_sensor_reporter {
+	int instance;
+	int type;
+	bool optional;
+};
+
+/**
+ * @struct senss_sensor_register_info
+ * @brief Sensor register information
+ *
+ * Can use this data structure to configure constant information when register
+ * a sensor instance
+ */
+struct senss_sensor_register_info {
+
+	/** Base constant information */
+	struct senss_sensor_info base_info;
+
+	/** Minimal sampling interval */
+	int minimal_interval;
+
+	/** API structure exposed by the sensor instance */
+	const void *api;
+
+	/** Reporters count*/
+	int reporters_count;
+	struct senss_sensor_reporter *reporters;
+};
+
+/**
+ * @struct senss_sensor
+ * @brief Internal sensor instance data structure.
+ *
+ * Each sensor instance will have it's unique data structure for storing all
+ * it's related information.
+ *
+ * Sensor subsystem core will enumerate and management of all these instance
+ * data structures, build report relationship model base on them, etc.
+ */
+struct senss_sensor {
+	/** Common Register info */
+	const struct senss_sensor_register_info *info;
+
+	/** Internal */
+
+	/** Sensor instance config information */
+	void *const config;
+
+	/** Sensor instance private data */
+	void *const data;
+
+	/** handle, updated in senss_init() */
+	const senss_sensor_handle_t handle;
+
+	/** reporter handles, updated in senss_init() */
+	int reporters_count;
+	const senss_sensor_handle_t *reporter_handles;
+
+	/** client handles, updated in senss_init() */
+	int clients_count;
+	const senss_sensor_handle_t *client_handles;
+};
+
+/**
+ * @brief Specific data section for sensor subsystem.
+ *
+ */
+#define __Z_SENSS_SENSOR__  \
+	__attribute__((used, section (".z_senss_sensor")))
+
+/**
+ * @brief Macro for register a sensor instance
+ *
+ * This macro will create some static data structures in specific data section:
+ * '.z_senss_sensor', \ref senss_init will enumerate all
+ * registered sensor instances from this data section.
+ *
+ * Based on \ref reporters_count value, if its zero, means no reporters
+ * configured, then sensor subsystem runtime will think its a physical sensor,
+ * otherwise is virtual sensor.
+ */
+#define SENSS_REGISTER_SENSOR(_inst, _name, _info, _config, _data)             \
+	static const typeof(_config) sensor_config_##_name##_inst;             \
+	static const typeof(_data) sensor_data_##_name##_instance;             \
+	static const struct senss_sensor                                       \
+	__Z_SENSS_SENSOR__ sensor##_name##_inst = {                            \
+		.info = _info,                                                 \
+		.config = &sensor_config_##_name##_instance,                   \
+		.data   = &sensor_data_##_name##_instance,                     \
+	};
```

Comment:
This very much looks like a `struct device`. How about making it look like a proper device? Reusing DEVICE_DT_INST_DEFINE(), having properties in the devicetree etc... that would also take care of the initialization stuff, multiple instances etc. The CDC-ACM driver went through something similar, I think that this would benefit from it as well.

Reply:
Yes,  it's looks like `struct device`.  There are two major reasons why at the first time not used `struct device` :  
* `senss` is user space framework for sensors,  the `senss` sensors (`physical` and `vritual`) in `senss` are pure software objects (`physical` senss sensors are interact with  underlying `sensor device drivers`.),  so, they're not real devices,  using `struct device` may be confusing. Seems `struct device` does support for application virtual device object abstraction,  but very few examples in zephyr code base. 
* `struct device`  devices will be initialized before `main` ,  that's good.  But `senss` sensor instances have reporting relationship during definition or registration,  such as `physical sensor A`  and `physical sensor B` reports to `virtual sensor D`,  `physical sensor B` also reports to `virtual sensor C`,  `virtual sensor C` is also another reporter of `virtual sensor D`.  We also want a  specific initialize process to initialize these sensors instances,  such as checking the reporting relationship,  if a dependency sensor type is missing etc,  and may need a specific init sequences according report dependencies.

As you said,  `struct device` is a very good fundamental concepts in zephyr  with many benefits, such as multiple instances, and devicetree,  which we also wanted in `senss` for sensors configuration,  so I will have a re-think about `struct device` .

Thanks a lot for point out his.

Reply:
Cool, the device stuff also has a way of handling initialization dependencies between devices, see https://github.com/zephyrproject-rtos/zephyr/blob/734c3454578d4db89b7a879b7339559b20c04c44/include/zephyr/device.h#L474

I think that it may sort out the initialization squence for free.

Reply:
Updated to `devicetree` and `struct device`, thanks!